### ISSUE TITLE
ACL replication error

### ISSUE THREAD

**Issue by jorgemarey** [hashicorp/consul#12295]
When filing a bug, please include the following headings if possible. Any example text in this template can be deleted.

#### Overview of the Issue

We are experiencing ACL Token replication problems between our DCs since upgrading from 1.9.5 to 1.10.3.
It seems that the error happen when a server has lost its leadership but gets elected again. In that case we don't see the ACL replication routines log messages.

#### Reproduction Steps

Not able to reproduce easily

### Consul info for both Client and Server

Consul v1.10.3

### Log Fragments

```
Feb  8 20:46:12 ctrl-5c33edfc-2b {"@level":"warn","@message":"failed to contact","@module":"agent.server.raft","@timestamp":"2022-02-08T20:46:12.145346Z","server-id":"e602e8f2-0599-e7f8-6cea-0cca235b17f4","time":758440418}
Feb  8 20:46:12 ctrl-5c33edfc-2b {"@level":"warn","@message":"failed to contact","@module":"agent.server.raft","@timestamp":"2022-02-08T20:46:12.145427Z","server-id":"76a22f32-068f-22a0-693d-4435a3edf459","time":802812278}
Feb  8 20:46:12 ctrl-5c33edfc-2b {"@level":"warn","@message":"failed to contact quorum of nodes, stepping down","@module":"agent.server.raft","@timestamp":"2022-02-08T20:46:12.145445Z"}
Feb  8 20:46:12 ctrl-5c33edfc-2b {"@level":"info","@message":"entering follower state","@module":"agent.server.raft","@timestamp":"2022-02-08T20:46:12.145485Z","follower":{},"leader":""}
Feb  8 20:46:12 ctrl-5c33edfc-2b {"@level":"info","@message":"aborting pipeline replication","@module":"agent.server.raft","@timestamp":"2022-02-08T20:46:12.146052Z","peer":{"Suffrage":0,"ID":"e602e8f2-0599-e7f8-6cea-0cca235b17f4","Address":"10.154.0.7:8300"}}
Feb  8 20:46:12 ctrl-5c33edfc-2b {"@level":"info","@message":"aborting pipeline replication","@module":"agent.server.raft","@timestamp":"2022-02-08T20:46:12.154979Z","peer":{"Suffrage":0,"ID":"76a22f32-068f-22a0-693d-4435a3edf459","Address":"10.154.0.12:8300"}}
Feb  8 20:46:12 ctrl-5c33edfc-2b {"@level":"error","@message":"Error when computing next state","@module":"agent.server.autopilot","@timestamp":"2022-02-08T20:46:12.155464Z","error":"cannot detect the current leader server id from its address: "}
Feb  8 20:46:12 ctrl-5c33edfc-2b {"@level":"info","@message":"cluster leadership lost","@module":"agent.server","@timestamp":"2022-02-08T20:46:12.155489Z"}
Feb  8 20:46:13 ctrl-5c33edfc-2b {"@level":"warn","@message":"heartbeat timeout reached, starting election","@module":"agent.server.raft","@timestamp":"2022-02-08T20:46:13.878313Z","last-leader":""}
Feb  8 20:46:13 ctrl-5c33edfc-2b {"@level":"info","@message":"entering candidate state","@module":"agent.server.raft","@timestamp":"2022-02-08T20:46:13.878404Z","node":{},"term":89571}
Feb  8 20:46:13 ctrl-5c33edfc-2b {"@level":"info","@message":"election won","@module":"agent.server.raft","@timestamp":"2022-02-08T20:46:13.888207Z","tally":2}
Feb  8 20:46:13 ctrl-5c33edfc-2b {"@level":"info","@message":"entering leader state","@module":"agent.server.raft","@timestamp":"2022-02-08T20:46:13.888249Z","leader":{}}
Feb  8 20:46:13 ctrl-5c33edfc-2b {"@level":"info","@message":"added peer, starting replication","@module":"agent.server.raft","@timestamp":"2022-02-08T20:46:13.888277Z","peer":"e602e8f2-0599-e7f8-6cea-0cca235b17f4"}
Feb  8 20:46:13 ctrl-5c33edfc-2b {"@level":"info","@message":"added peer, starting replication","@module":"agent.server.raft","@timestamp":"2022-02-08T20:46:13.888296Z","peer":"76a22f32-068f-22a0-693d-4435a3edf459"}
Feb  8 20:46:13 ctrl-5c33edfc-2b {"@level":"info","@message":"cluster leadership acquired","@module":"agent.server","@timestamp":"2022-02-08T20:46:13.888491Z"}
Feb  8 20:46:13 ctrl-5c33edfc-2b {"@level":"info","@message":"New leader elected","@module":"agent.server","@timestamp":"2022-02-08T20:46:13.888698Z","payload":"ctrl-5c33edfc-2b"}
Feb  8 20:46:13 ctrl-5c33edfc-2b {"@level":"info","@message":"pipelining replication","@module":"agent.server.raft","@timestamp":"2022-02-08T20:46:13.889246Z","peer":{"Suffrage":0,"ID":"e602e8f2-0599-e7f8-6cea-0cca235b17f4","Address":"10.154.0.7:8300"}}
Feb  8 20:46:13 ctrl-5c33edfc-2b {"@level":"info","@message":"pipelining replication","@module":"agent.server.raft","@timestamp":"2022-02-08T20:46:13.889336Z","peer":{"Suffrage":0,"ID":"76a22f32-068f-22a0-693d-4435a3edf459","Address":"10.154.0.12:8300"}}
Feb  8 20:46:13 ctrl-5c33edfc-2b {"@level":"info","@message":"started routine","@module":"agent.leader","@timestamp":"2022-02-08T20:46:13.894833Z","routine":"acl token reaping"}
Feb  8 20:46:13 ctrl-5c33edfc-2b {"@level":"info","@message":"started routine","@module":"agent.leader","@timestamp":"2022-02-08T20:46:13.895703Z","routine":"config entry replication"}
Feb  8 20:46:13 ctrl-5c33edfc-2b {"@level":"info","@message":"started routine","@module":"agent.leader","@timestamp":"2022-02-08T20:46:13.895737Z","routine":"federation state replication"}
Feb  8 20:46:13 ctrl-5c33edfc-2b {"@level":"info","@message":"started routine","@module":"agent.leader","@timestamp":"2022-02-08T20:46:13.895753Z","routine":"federation state anti-entropy"}
Feb  8 20:46:14 ctrl-5c33edfc-2b {"@level":"info","@message":"initialized secondary datacenter CA with provider","@module":"agent.server.connect","@timestamp":"2022-02-08T20:46:14.057474Z","provider":"consul"}
Feb  8 20:46:14 ctrl-5c33edfc-2b {"@level":"info","@message":"started routine","@module":"agent.leader","@timestamp":"2022-02-08T20:46:14.057496Z","routine":"secondary CA roots watch"}
Feb  8 20:46:14 ctrl-5c33edfc-2b {"@level":"info","@message":"started routine","@module":"agent.leader","@timestamp":"2022-02-08T20:46:14.057510Z","routine":"intermediate cert renew watch"}
Feb  8 20:46:14 ctrl-5c33edfc-2b {"@level":"info","@message":"started routine","@module":"agent.leader","@timestamp":"2022-02-08T20:46:14.057525Z","routine":"CA root pruning"}
```

In a normal restart the servers starts the following routines:

```
Feb  9 02:15:48 ctrl-aa06beda-1b {"@level":"info","@message":"started routine","@module":"agent.leader","@timestamp":"2022-02-09T02:15:48.052292Z","routine":"ACL policy replication"}
Feb  9 02:15:48 ctrl-aa06beda-1b {"@level":"info","@message":"started routine","@module":"agent.leader","@timestamp":"2022-02-09T02:15:48.052323Z","routine":"ACL role replication"}
Feb  9 02:15:48 ctrl-aa06beda-1b {"@level":"info","@message":"started routine","@module":"agent.leader","@timestamp":"2022-02-09T02:15:48.052338Z","routine":"ACL token replication"}
Feb  9 02:15:48 ctrl-aa06beda-1b {"@level":"info","@message":"started routine","@module":"agent.leader","@timestamp":"2022-02-09T02:15:48.052351Z","routine":"acl token reaping"}
Feb  9 02:15:48 ctrl-aa06beda-1b {"@level":"info","@message":"started ACL Policy replication","@module":"agent.server.replication.acl.policy","@timestamp":"2022-02-09T02:15:48.052644Z"}
Feb  9 02:15:48 ctrl-aa06beda-1b {"@level":"info","@message":"started ACL Role replication","@module":"agent.server.replication.acl.role","@timestamp":"2022-02-09T02:15:48.052859Z"}
Feb  9 02:15:48 ctrl-aa06beda-1b {"@level":"info","@message":"started ACL Token replication","@module":"agent.server.replication.acl.token","@timestamp":"2022-02-09T02:15:48.053018Z"}
Feb  9 02:15:48 ctrl-aa06beda-1b {"@level":"info","@message":"started routine","@module":"agent.leader","@timestamp":"2022-02-09T02:15:48.066174Z","routine":"config entry replication"}
Feb  9 02:15:48 ctrl-aa06beda-1b {"@level":"info","@message":"started routine","@module":"agent.leader","@timestamp":"2022-02-09T02:15:48.066222Z","routine":"federation state replication"}
Feb  9 02:15:48 ctrl-aa06beda-1b {"@level":"info","@message":"started routine","@module":"agent.leader","@timestamp":"2022-02-09T02:15:48.066245Z","routine":"federation state anti-entropy"}
Feb  9 02:15:48 ctrl-aa06beda-1b {"@level":"info","@message":"started routine","@module":"agent.leader","@timestamp":"2022-02-09T02:15:48.519228Z","routine":"secondary CA roots watch"}
Feb  9 02:15:48 ctrl-aa06beda-1b {"@level":"info","@message":"started routine","@module":"agent.leader","@timestamp":"2022-02-09T02:15:48.519259Z","routine":"intermediate cert renew watch"}
Feb  9 02:15:48 ctrl-aa06beda-1b {"@level":"info","@message":"started routine","@module":"agent.leader","@timestamp":"2022-02-09T02:15:48.519311Z","routine":"CA root pruning"}
```

I was looking at the code and maybe the problem is related to the following check, that when a leader regains leadership soon after lost it the ACL replication routines are not started.

https://github.com/hashicorp/consul/blob/c976ffd2d4d8952404e3fc4fa784357bd1f7c36b/agent/consul/leader.go#L853-L858

**Comment by Amier3** [comment]
Hey @jorgemarey 

Thanks for bringing this to our attention, we'll look into this and let you know if we need some further info/logs 👍

**Comment by weichuliu** [comment]
We experience exactly the same issue in our deployment (on 1.10.6+ent), and this is what we found.

#### TL;DR

The ACL replication stopping issue is triggered *only when the leadership is lost and regained immediately after by the same server*.

The root cause is what @jorgemarey points out -- When the new ACL routines try to start, new routines still in running state (because ACL fetching is a blocking call). Then the ACL replication routines won't be restart until the next leader election.

The ACL bug has been there, also in 1.9. But I believe 1.10 introduced some new behavior that makes the "lost and regain leadership" occurring much more frequently, which exposes this ACL stop issue.

#### Investigation

Here is the leader log with TRACE during a "lost and regain leadership".

```
2022-02-25T03:08:26.237Z [DEBUG] agent.server.replication.acl.token: finished fetching acls: amount=252
2022-02-25T03:08:26.238Z [DEBUG] agent.server.replication.acl.token: acl replication: local=252 remote=252
2022-02-25T03:08:26.238Z [DEBUG] agent.server.replication.acl.token: acl replication: deletions=0 updates=0
2022-02-25T03:08:26.238Z [DEBUG] agent.server.replication.acl.token: ACL replication completed through remote index: index=2011524
2022-02-25T03:08:31.627Z [DEBUG] agent.server.replication.acl.policy: finished fetching acls: amount=714
2022-02-25T03:08:31.627Z [DEBUG] agent.server.replication.acl.policy: acl replication: local=714 remote=714
2022-02-25T03:08:31.627Z [DEBUG] agent.server.replication.acl.policy: acl replication: deletions=0 updates=0
2022-02-25T03:08:31.627Z [DEBUG] agent.server.replication.acl.policy: ACL replication completed through remote index: index=1578781
2022-02-25T03:12:08.419Z [DEBUG] agent.server.replication.acl.role: finished fetching acls: amount=0
2022-02-25T03:12:08.419Z [DEBUG] agent.server.replication.acl.role: acl replication: local=0 remote=0
2022-02-25T03:12:08.419Z [DEBUG] agent.server.replication.acl.role: acl replication: deletions=0 updates=0
2022-02-25T03:12:08.419Z [DEBUG] agent.server.replication.acl.role: ACL replication completed through remote index: index=1
2022-02-25T03:13:38.834Z [WARN]  agent.server.raft: failed to contact: server-id=80fea010-60c6-7219-0e4c-38263612172c time=903.198574ms
2022-02-25T03:13:38.834Z [WARN]  agent.server.raft: failed to contact: server-id=fcd1f4ec-b3e7-3099-590d-7880119e3f0b time=903.610762ms
2022-02-25T03:13:38.834Z [WARN]  agent.server.raft: failed to contact quorum of nodes, stepping down
2022-02-25T03:13:38.834Z [INFO]  agent.server.raft: entering follower state: follower="Node at 10.117.42.237:8300 [Follower]" leader=
2022-02-25T03:13:38.835Z [INFO]  agent.server.raft: aborting pipeline replication: peer="{Voter fcd1f4ec-b3e7-3099-590d-7880119e3f0b 10.117.106.248:8300}"
2022-02-25T03:13:38.835Z [DEBUG] agent.leader: stopping routine: routine="namespace replication"
2022-02-25T03:13:38.835Z [DEBUG] agent.leader: stopping routine: routine="namespace deferred deletion"
2022-02-25T03:13:38.835Z [DEBUG] agent.leader: stopping routine: routine="partition deferred deletion"
2022-02-25T03:13:38.835Z [DEBUG] agent.leader: stopping routine: routine="federation state anti-entropy"
2022-02-25T03:13:38.835Z [DEBUG] agent.leader: stopping routine: routine="federation state replication"
2022-02-25T03:13:38.835Z [DEBUG] agent.leader: stopping routine: routine="config entry replication"
2022-02-25T03:13:38.835Z [DEBUG] agent.leader: stopping routine: routine="acl token reaping"
2022-02-25T03:13:38.835Z [DEBUG] agent.leader: stopped routine: routine="namespace deferred deletion"
2022-02-25T03:13:38.835Z [DEBUG] agent.leader: stopped routine: routine="partition deferred deletion"
2022-02-25T03:13:38.835Z [DEBUG] agent.leader: stopped routine: routine="acl token reaping"
2022-02-25T03:13:38.835Z [DEBUG] agent.server.autopilot: state update routine is now stopped
2022-02-25T03:13:38.835Z [INFO]  agent.server.raft: aborting pipeline replication: peer="{Voter 80fea010-60c6-7219-0e4c-38263612172c 10.117.170.233:8300}"
2022-02-25T03:13:39.952Z [WARN]  agent.server.raft: heartbeat timeout reached, starting election: last-leader=
2022-02-25T03:13:39.952Z [INFO]  agent.server.raft: entering candidate state: node="Node at 10.117.42.237:8300 [Candidate]" term=3620
2022-02-25T03:13:39.957Z [DEBUG] agent.server.raft: votes: needed=2
2022-02-25T03:13:39.957Z [DEBUG] agent.server.raft: vote granted: from=87ff1d3f-6d41-a626-a8bb-ae81fc2d61c1 term=3620 tally=1
2022-02-25T03:13:39.960Z [DEBUG] agent.server.raft: vote granted: from=80fea010-60c6-7219-0e4c-38263612172c term=3620 tally=2
2022-02-25T03:13:39.960Z [INFO]  agent.server.raft: election won: tally=2
2022-02-25T03:13:39.960Z [INFO]  agent.server.raft: entering leader state: leader="Node at 10.117.42.237:8300 [Leader]"
2022-02-25T03:13:39.960Z [INFO]  agent.server.raft: added peer, starting replication: peer=80fea010-60c6-7219-0e4c-38263612172c
2022-02-25T03:13:39.960Z [INFO]  agent.server.raft: added peer, starting replication: peer=fcd1f4ec-b3e7-3099-590d-7880119e3f0b
2022-02-25T03:13:39.961Z [INFO]  agent.server.raft: pipelining replication: peer="{Voter 80fea010-60c6-7219-0e4c-38263612172c 10.117.170.233:8300}"
2022-02-25T03:13:39.963Z [INFO]  agent.server.raft: pipelining replication: peer="{Voter fcd1f4ec-b3e7-3099-590d-7880119e3f0b 10.117.106.248:8300}"
2022-02-25T03:13:39.967Z [INFO]  agent.leader: started routine: routine="acl token reaping"
2022-02-25T03:13:39.967Z [INFO]  agent.leader: started routine: routine="namespace deferred deletion"
2022-02-25T03:13:39.967Z [INFO]  agent.leader: started routine: routine="partition deferred deletion"
2022-02-25T03:13:39.967Z [INFO]  agent.leader: started routine: routine="namespace replication"
2022-02-25T03:13:39.969Z [INFO]  agent.leader: started routine: routine="config entry replication"
2022-02-25T03:13:39.969Z [INFO]  agent.leader: started routine: routine="federation state replication"
2022-02-25T03:13:39.969Z [INFO]  agent.leader: started routine: routine="federation state anti-entropy"
2022-02-25T03:13:39.969Z [DEBUG] agent.server.autopilot: state update routine is now running
2022-02-25T03:13:43.035Z [DEBUG] agent.server.replication.acl.token: finished fetching acls: amount=252
2022-02-25T03:13:43.035Z [DEBUG] agent.leader: stopped routine: routine="ACL token replication"
2022-02-25T03:13:45.412Z [DEBUG] agent.server.replication.acl.policy: finished fetching acls: amount=714
2022-02-25T03:13:45.412Z [DEBUG] agent.leader: stopped routine: routine="ACL policy replication"
2022-02-25T03:15:54.773Z [DEBUG] agent.leader: stopped routine: routine="config entry replication"
2022-02-25T03:16:38.502Z [DEBUG] agent.leader: stopped routine: routine="federation state anti-entropy"
2022-02-25T03:17:05.626Z [DEBUG] agent.leader: stopped routine: routine="federation state replication"
2022-02-25T03:17:13.897Z [DEBUG] agent.server.replication.acl.role: finished fetching acls: amount=0
2022-02-25T03:17:13.897Z [DEBUG] agent.leader: stopped routine: routine="ACL role replication"
2022-02-25T03:17:59.952Z [DEBUG] agent.leader: stopped routine: routine="namespace replication"
```

When the leadership is lost, [revokeLeadership](https://github.com/hashicorp/consul/blob/f3b9dc5bcaeb74199fd7b0888954b402433844c8/agent/consul/leader.go#L372) will be called. When leadership is regained, [establishLeadership](https://github.com/hashicorp/consul/blob/f3b9dc5bcaeb74199fd7b0888954b402433844c8/agent/consul/leader.go#L306) will be called.

In `establishLeadership`, the [initializeACLs](https://github.com/hashicorp/consul/blob/f3b9dc5bcaeb74199fd7b0888954b402433844c8/agent/consul/leader.go#L311) will be called, and will [stopACLReplication then startACLReplication](https://github.com/hashicorp/consul/blob/f3b9dc5bcaeb74199fd7b0888954b402433844c8/agent/consul/leader.go#L687-L692).

In the `startACLReplication`, if there are already [ACL replication routines running](https://github.com/hashicorp/consul/blob/f3b9dc5bcaeb74199fd7b0888954b402433844c8/agent/consul/leader.go#L856). New routines won't be created.

However, the ACL routines are blocking calls. The routines don't stop immediately after `s.leaderRoutineManager.Stop` are called, they stop when the fetching RPC is finished.

This is clearly described in the [comment](https://github.com/hashicorp/consul/blob/f3b9dc5bcaeb74199fd7b0888954b402433844c8/agent/consul/leader.go#L874-L875):

```
// This function is only intended to be run as a managed go routine, it will block until
// the context passed in indicates that it should exit.
```

Indeed, our logs prove that the ACL fetching call lasted a while after re-election:

```
2022-02-25T03:13:39.969Z [DEBUG] agent.server.autopilot: state update routine is now running

2022-02-25T03:13:43.035Z [DEBUG] agent.server.replication.acl.token: finished fetching acls: amount=252
2022-02-25T03:13:43.035Z [DEBUG] agent.leader: stopped routine: routine="ACL token replication"
2022-02-25T03:13:45.412Z [DEBUG] agent.server.replication.acl.policy: finished fetching acls: amount=714
2022-02-25T03:13:45.412Z [DEBUG] agent.leader: stopped routine: routine="ACL policy replication"
2022-02-25T03:17:13.897Z [DEBUG] agent.server.replication.acl.role: finished fetching acls: amount=0
2022-02-25T03:17:13.897Z [DEBUG] agent.leader: stopped routine: routine="ACL role replication"
```

The `establishLeadership` ran at around 13:39.9, but the three routines stopped 5s to 4min later.

The `finished fetching acls` comes from [replicateACLType](https://github.com/hashicorp/consul/blob/f3b9dc5bcaeb74199fd7b0888954b402433844c8/agent/consul/acl_replication.go#L376-L383). 

The actual ACL update which generates logs like `acl replication: local=714 remote=714` is not seen here, meaning that the code realized the context is canceled, then exit.

#### How to fix

In the [routine.go](https://github.com/hashicorp/consul/blob/f3b9dc5bcaeb74199fd7b0888954b402433844c8/lib/routine/routine.go#L86), the routineTracker uses an extra stoppedCh to store the routine close status. The routine.running() only checks this channel to tell whether a routine isRunning.

However, this channel is only closed after the `execute` is finished. When the routine has blocking call such as ACL FetchRemote, the context can be already canceled for a while however the stoppedCh is still open.

As we can see in the logs, the ACL role routine stopped 4 minutes after Stop is issued.

Actually, the routine's ctx's Done channel should be sufficient for telling the running state of the routine, and this stoppedCh can be removed to avoid providing false information
